File: Java/Hard/Accepted/24-game_4.java
class Solution {
    public boolean judgePoint24(int[] cards) {
        
        List<Double> list = new ArrayList<>();
        for (int card : cards) {
            list.add((double) card);
        }
        return judgePoint24(list);
    }

    private boolean judgePoint24(List<Double> list) {
        if (list.size() == 0) {
            return false;
        }
        if (list.size() == 1) {
            return Math.abs(list.get(0) - 24) < 1e-6;
        }
        int size = list.size();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (i != j) {
                    List<Double> list2 = new ArrayList<>();
                    for (int k = 0; k < size; k++) {
                        if (k != i && k != j) {
                            list2.add(list.get(k));
                        }
                    }
                    for (int k = 0; k < 4; k++) {
                        if (k < 2 && i > j) {
                            continue;
                        }
                        if (k == 0) {
                            list2.add(list.get(i) + list.get(j));
                        } else if (k == 1) {
                            list2.add(list.get(i) * list.get(j));
                        } else if (k == 2) {
                            list2.add(list.get(i) - list.get(j));
                        } else {
                            if (Math.abs(list.get(j)) < 1e-6) {
                                continue;
                            }
                            list2.add(list.get(i) / list.get(j));
                        }
                        if (judgePoint24(list2)) {
                            return true;
                        }
                        list2.remove(list2.size() - 1);
                    }
                }
            }
        }
        return false;
    }
}

File: Python3/Hard/Accepted/24-game_4.py
class Solution:
    def judgePoint24(self, cards: List[int]) -> bool:
        
        def dfs(cards):
            if len(cards) == 1:
                return abs(cards[0] - 24) < 1e-6
            for i in range(len(cards)):
                for j in range(len(cards)):
                    if i != j:
                        new_cards = [cards[k] for k in range(len(cards)) if k != i and k != j]
                        for op in range(6):
                            if op == 0:
                                new_cards.append(cards[i] + cards[j])
                            elif op == 1:
                                new_cards.append(cards[i] * cards[j])
                            elif op == 2:
                                new_cards.append(cards[i] - cards[j])
                            elif op == 3:
                                if cards[j] != 0:
                                    new_cards.append(cards[i] / cards[j])
                                else:
                                    continue
                            elif op == 4:
                                if cards[i] != 0:
                                    new_cards.append(cards[j] / cards[i])
                                else:
                                    continue
                            else:
                                continue
                            if dfs(new_cards):
                                return True
                            new_cards.pop()
            return False
        return dfs(cards)